TESTS/mbed_hal/trng/main.cpp

```diff
@@ -0,0 +1,230 @@
+* WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+/*
+* The test is based on the assumption that trng will generate random data, random so
+* there will not be any similar patterns in it, that kind of data will be impossible to
+* compress, if compression will occur the test will result in failure.
+*
+* The test is composed out of three parts:
+* the first, generate a trng buffer and try to compress it, at the end of first part
+* we will reset the device.
+* In second part we will generate a trng buffer with a different buffer size and try to
+* compress it.
+* In the third part we will again generate a trng buffer to see that the same trng output
+* is not generated as the stored trng buffer from part one (before reseting), the new trng data will
+* be concatenated to the trng data from the first part and then try to compress it
+* together, if there are similar patterns the compression will succeed.
+*
+* We need to store and load the first part data before and after reset, the mechanism
+* we will use is the mbed greentea platform for sending and receving the data from the device
+* to the host running the test and back, the problem with this mechanism is that it doesn't
+* handle well certain characters, especially non ASCII ones, so we use the base64 algorithm
+* to ensure all characters will be transmitted correctly.
+*/
+
+#include "greentea-client/test_env.h"
+#include "unity/unity.h"
+#include "utest/utest.h"
+#include "hal/trng_api.h"
+#include "base64b.h"
+#include "pithy.h"
+#include <stdio.h>
+
+#if !DEVICE_TRNG
+#error [NOT_SUPPORTED] TRNG API not supported for this target
+#endif
+
+#define MSG_VALUE_DUMMY                 "0"
+#define MSG_VALUE_LEN                   64
+#define MSG_KEY_LEN                     32
+
+#define BUFFER_LEN                      (MSG_VALUE_LEN/2)
+
+#define MSG_TRNG_READY                  "ready"
+#define MSG_TRNG_BUFFER                 "buffer"
+#define MSG_TRNG_EXIT                   "exit"
+
+#define MSG_TRNG_TEST_STEP1             "check_step1"
+#define MSG_TRNG_TEST_STEP2             "check_step2"
+#define MSG_TRNG_TEST_SUITE_ENDED       "Test_suite_ended"
+
+#define RESULT_SUCCESS                  0
+
+using namespace utest::v1;
+
+static int fill_buffer_trng(uint8_t *buffer, trng_t *trng_obj, size_t trng_len)
+{
+    size_t temp_size = 0, output_length = 0;
+    int trng_res = 0;
+    uint8_t *temp_in_buf = buffer;
+
+    trng_init(trng_obj);
+    memset(buffer, 0, BUFFER_LEN);
+
+    while (true) {
+        trng_res = trng_get_bytes(trng_obj, temp_in_buf, trng_len - temp_size, &output_length);
+        TEST_ASSERT_EQUAL_INT_MESSAGE(0, trng_res, "trng_get_bytes error!");
+        temp_size += output_length;
+        temp_in_buf += output_length;
+        trng_len -= output_length;
+        if (temp_size >= trng_len) {
```

Comment:
* Caller requests 32 random bytes: `trng_len` = 32; `temp_size` = 0
* Implementation returns 16 random bytes: `output_length`= 16
* `temp_size += output_length;`: `temp_size` = 16
* `trng_len -= output_length;`: `trng_len` = 16
* `if (temp_size >= trng_len) {`: evaluates to true
* `break;`: function exits successfully but last 16 bytes of the buffer are still set to 0